<!doctype html><html lang=en><head><title>Mastering Python's Exceptional Complexity</title>
<meta charset=UTF-8><meta name=viewport content="width=device-width,initial-scale=1"><meta http-equiv=X-UA-Compatible content="ie=edge"><link rel=stylesheet href=/application.ffc4c7b40a84a6cc07476bf220dd79432177a593771276da69cb703d50fcac34.css integrity="sha256-/8THtAqEpswHR2vyIN15QyF3pZN3EnbaactwPVD8rDQ="><link rel=icon type=image/png href=/images/site/favicon_hu14458373468778497799.png><meta property="og:url" content="https://carlvinjerry.github.io/posts/programming/code-sustainability/error-handling/python-exceptions/"><meta property="og:site_name" content="RΞCURSION"><meta property="og:title" content="Mastering Python's Exceptional Complexity"><meta property="og:description" content="Mastering Python's Exceptional Complexity for Sustainable Error Handling"><meta property="og:locale" content="en"><meta property="og:type" content="article"><meta property="article:section" content="posts"><meta property="article:published_time" content="2023-09-14T06:00:20+06:00"><meta property="article:modified_time" content="2025-08-18T17:45:06+03:00"><meta property="article:tag" content="Python"><meta property="article:tag" content="Exceptions"><meta property="article:tag" content="Code Sustainability"><meta name=twitter:card content="summary"><meta name=twitter:title content="Mastering Python's Exceptional Complexity"><meta name=twitter:description content="Mastering Python's Exceptional Complexity for Sustainable Error Handling"><meta name=description content="Mastering Python's Exceptional Complexity for Sustainable Error Handling"><script async src="https://www.googletagmanager.com/gtag/js?id=462784605"></script><script>var dnt,doNotTrack=!1;if(null&&(dnt=navigator.doNotTrack||window.doNotTrack||navigator.msDoNotTrack,doNotTrack=dnt=="1"||dnt=="yes"),!doNotTrack){window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments)}gtag("js",new Date),gtag("config","462784605")}</script><script src=https://cdn.counter.dev/script.js data-id=CarlvinJerry data-utcoffset=1></script><script data-goatcounter=https://carlvinjerry.goatcounter.com/count async src=//gc.zgo.at/count.js></script><script integrity="sha256-DO4ugzEwhTW1Id1UIWn0gUJWaebCYOypeTit6LW4QB4=">let theme=localStorage.getItem("theme-scheme")||localStorage.getItem("darkmode:color-scheme")||"light";theme==="system"&&(window.matchMedia&&window.matchMedia("(prefers-color-scheme: dark)").matches?theme="dark":theme="light"),document.documentElement.setAttribute("data-theme",theme)</script></head><body class="type-posts kind-page" data-bs-spy=scroll data-bs-target=#TableOfContents data-bs-offset=80><div class="container-fluid bg-secondary wrapper"><nav class="navbar navbar-expand-xl top-navbar shadow" id=top-navbar><div class=container><button class="navbar-toggler navbar-light" id=sidebar-toggler type=button>
<i data-feather=sidebar></i>
</button>
<a class=navbar-brand href=/><img src=/images/site/main-logo_hu1754343916303502928.png id=logo alt=Logo>
RΞCURSION</a>
<button class="navbar-toggler navbar-light" id=navbar-toggler type=button data-bs-toggle=collapse data-bs-target=#top-nav-items aria-label=menu>
<i data-feather=menu></i></button><div class="collapse navbar-collapse dynamic-navbar" id=top-nav-items><ul class="nav navbar-nav ms-auto"><li class=nav-item><a class=nav-link href=/#home>Home</a></li><li class=nav-item><a class=nav-link href=/#about>About</a></li><li class=nav-item><a class=nav-link href=/#recent-posts>Recent Posts</a></li><li class=nav-item><a class=nav-link href=/#projects>Projects</a></li><div id=top-navbar-divider></div><li class=nav-item><a class=nav-link id=blog-link href=/posts>Posts</a></li><li class=nav-item><a class=nav-link id=note-link href=/notes>Notes</a></li><li class=nav-item><a class=nav-link href=https://toha-guides.netlify.app/posts/>Docs</a></li><li class="nav-item dropdown"><a class="nav-link dropdown-toggle" href=# id=themeSelector role=button data-bs-toggle=dropdown aria-haspopup=true aria-expanded=false><img id=navbar-theme-icon-svg class=theme-icon src=/icons/moon-svgrepo-com.svg width=20 alt="Dark Theme"></a><div id=themeMenu class="dropdown-menu dropdown-menu-icons-only" aria-labelledby=themeSelector><a class="dropdown-item nav-link" href=# data-scheme=light><img class=theme-icon src=/icons/sun-svgrepo-com.svg width=20 alt="Light Theme">
</a><a class="dropdown-item nav-link" href=# data-scheme=dark><img class=theme-icon src=/icons/moon-svgrepo-com.svg width=20 alt="Dark Theme">
</a><a class="dropdown-item nav-link" href=# data-scheme=system><img class=theme-icon src=/icons/computer-svgrepo-com.svg width=20 alt="System Theme"></a></div></li></ul></div></div><img src=/images/site/main-logo_hu1754343916303502928.png class=d-none id=main-logo alt=Logo>
<img src=/images/site/inverted-logo_hu14458373468778497799.png class=d-none id=inverted-logo alt="Inverted Logo"></nav><section class=sidebar-section id=sidebar-section><div class=sidebar-holder><div class=sidebar id=sidebar><form class=mx-auto method=get action=/search><input type=text name=keyword placeholder=Search data-search id=search-box></form><div class=sidebar-tree><ul class=tree id=tree><li id=list-heading><a href=/posts/ data-filter=all>Posts</a></li><div class=subtree><li><a class=list-link href=/posts/introduction/ title=Introduction>Introduction</a></li><li><i data-feather=minus-circle></i><a class="active list-link" href=/posts/programming/> Programming</a><ul class=active><li><i data-feather=minus-circle></i><a class="active list-link" href=/posts/programming/code-sustainability/> Code Sustainability</a><ul class=active><li><i data-feather=minus-circle></i><a class="active list-link" href=/posts/programming/code-sustainability/error-handling/> Error Handling</a><ul class=active><li><a class="active list-link" href=/posts/programming/code-sustainability/error-handling/python-exceptions/ title="Exception Handling in Python">Exception Handling in Python</a></li></ul></li><li><i data-feather=plus-circle></i><a class=list-link href=/posts/programming/code-sustainability/succint-fsharp/> Succint FSharp</a><ul><li><a class=list-link href=/posts/programming/code-sustainability/succint-fsharp/currying/ title="Currying and Partial Application">Currying and Partial Application</a></li><li><a class=list-link href=/posts/programming/code-sustainability/succint-fsharp/working-with-paket/ title="A Better Way to Manage Dependencies in F#">A Better Way to Manage Dependencies in F#</a></li><li><a class=list-link href=/posts/programming/code-sustainability/succint-fsharp/memoization/ title="Memoization in F#">Memoization in F#</a></li><li><a class=list-link href=/posts/programming/code-sustainability/succint-fsharp/pattern-matching/ title="Pattern Matching in F#">Pattern Matching in F#</a></li><li><a class=list-link href=/posts/programming/code-sustainability/succint-fsharp/fs-option-type/ title="Understanding the F# Option Type">Understanding the F# Option Type</a></li></ul></li></ul></li><li><i data-feather=plus-circle></i><a class=list-link href=/posts/programming/functional-programming/> Functional Programming</a><ul><li><a class=list-link href=/posts/programming/functional-programming/discriminated-unions/ title="Discriminated Unions in F#">Discriminated Unions in F#</a></li><li><a class=list-link href=/posts/programming/functional-programming/functional-programming/ title="Functional Programming Nuances">Functional Programming Nuances</a></li><li><a class=list-link href=/posts/programming/functional-programming/monads/ title="Functors and Monads in F#">Functors and Monads in F#</a></li></ul></li><li><i data-feather=plus-circle></i><a class=list-link href=/posts/programming/interoperability/> Interoperability</a><ul><li><a class=list-link href=/posts/programming/interoperability/rdotnet-interop/ title="R.NET Integration in .NET">R.NET Integration in .NET</a></li></ul></li><li><i data-feather=plus-circle></i><a class=list-link href=/posts/programming/languages/> Languages</a><ul><li><i data-feather=plus-circle></i><a class=list-link href=/posts/programming/languages/fsharp/> fsharp</a><ul><li><a class=list-link href=/posts/programming/languages/fsharp/fsharp-9-review/ title=fsharp-9-review>fsharp-9-review</a></li></ul></li></ul></li></ul></li><li><i data-feather=plus-circle></i><a class=list-link href=/posts/quantitative-finance/> Quantitative-Finance</a><ul><li><a class=list-link href=/posts/quantitative-finance/intro-to-sdes/ title="Introduction to SDEs">Introduction to SDEs</a></li><li><a class=list-link href=/posts/quantitative-finance/portfolio-risk-attribution/ title="Portfolio Risk Attribution">Portfolio Risk Attribution</a></li><li><a class=list-link href=/posts/quantitative-finance/moving-averages/ title="Simple Moving Average">Simple Moving Average</a></li></ul></li></div></ul></div></div></div></section><section class=content-section id=content-section><div class=content><div class="container p-0 read-area"><div class="hero-area col-sm-12" id=hero-area style=background-image:url(/posts/programming/code-sustainability/error-handling/python-exceptions/errors.jpg)></div><div class=page-content><div class="author-profile ms-auto align-self-lg-center"><img class=rounded-circle src=/images/author/j5_hu8605642293593435631.jpg alt="Author Image"><h5 class=author-name>Carlvin Jerry Mwange</h5><p class=text-muted>Thursday, September 14, 2023 | 8 minutes</p></div><div class=title><h1>Mastering Python's Exceptional Complexity</h1></div><div class=tags><ul style=padding-left:0><li class=rounded><a href=/tags/python/ class="btn btn-sm btn-info">Python</a></li><li class=rounded><a href=/tags/exceptions/ class="btn btn-sm btn-info">Exceptions</a></li><li class=rounded><a href=/tags/code-sustainability/ class="btn btn-sm btn-info">Code Sustainability</a></li></ul></div><div class=post-content id=post-content><h4 id=introduction>Introduction</h4><p>Exception handling is a fundamental pillar of programming, and Python in particular, excels in this area given its robust and versatile error management system. A program terminates immediately it encounters an error, typically taking one of two forms: <strong>syntax errors</strong> and <strong>exceptions</strong>. In this article, we&rsquo;ll demystify exceptions, distinguishing them from syntax errors, and explore the art of effectively managing them. Finally, we will look at some advanced techniques coupled with best practices for gracefully handling exceptions.</p><br><h5 id=exceptions-versus-syntax-errors>Exceptions versus Syntax Errors</h5><p><a href=https://realpython.com/invalid-syntax-python/ target=_blank>Syntax errors</a> occur during the <em><strong>parsing phase</strong></em> before code execution. The interpreter detects syntax errors whenever the code violates the rules and structure of the Python language, leading to a complete halt of the program. Observe the following examples:</p><ul><li>Missing a colon after a def statement:</li></ul><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#75715e>#Code with missing collon</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>def</span> <span style=color:#a6e22e>my_function</span>(a , b)  <span style=color:#75715e>#SyntaxError:unexpectedEOFwhileparsing</span>
</span></span><span style=display:flex><span> print(a <span style=color:#f92672>+</span> b)
</span></span></code></pre></div><br><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span> File <span style=color:#e6db74>&#34;&lt;ipython-input-2-b7d4e84eafb5&gt;&#34;</span>, line <span style=color:#ae81ff>2</span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>def</span> <span style=color:#a6e22e>my_function</span>(a , b)  <span style=color:#75715e># SyntaxError: unexpected EOF while parsing</span>
</span></span><span style=display:flex><span>                       <span style=color:#f92672>^</span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>SyntaxError</span>: expected <span style=color:#e6db74>&#39;:&#39;</span>
</span></span></code></pre></div><br><p>The arrow in the our output indicates exactly where the parser ran into the <strong>syntax error</strong>. Fixing the code yields a different result. As an example, testing the function with any two integers gives us their sum:</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#75715e>#Fixed code</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>def</span> <span style=color:#a6e22e>my_function</span>(a , b):<span style=color:#75715e># No error</span>
</span></span><span style=display:flex><span> print( a<span style=color:#f92672>+</span>b )
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#75715e>#Test</span>
</span></span><span style=display:flex><span>my_function(<span style=color:#ae81ff>5</span>,<span style=color:#ae81ff>6</span>)
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#f92672>&gt;&gt;</span><span style=color:#ae81ff>11</span>
</span></span></code></pre></div><br><ul><li>Using an undefined variable:</li></ul><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>print(undefined_variable)  <span style=color:#75715e># NameError: name &#39;undefined_variable&#39; is not defined</span>
</span></span></code></pre></div><br><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#f92672>---------------------------------------------------------------------------</span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>NameError</span>                                 Traceback (most recent call last)
</span></span><span style=display:flex><span><span style=color:#f92672>&lt;</span>ipython<span style=color:#f92672>-</span>input<span style=color:#f92672>-</span><span style=color:#ae81ff>12</span><span style=color:#f92672>-</span>b3b3b04032c5<span style=color:#f92672>&gt;</span> <span style=color:#f92672>in</span> <span style=color:#f92672>&lt;</span>cell line: <span style=color:#ae81ff>1</span><span style=color:#f92672>&gt;</span>()
</span></span><span style=display:flex><span><span style=color:#f92672>----&gt;</span> <span style=color:#ae81ff>1</span> print(undefined_variable)
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>NameError</span>: name <span style=color:#e6db74>&#39;undefined_variable&#39;</span> <span style=color:#f92672>is</span> <span style=color:#f92672>not</span> defined
</span></span></code></pre></div><br><ul><li>Mismatched parentheses:</li></ul><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>print(<span style=color:#e6db74>&#34;Hello, World!&#34;</span>  <span style=color:#75715e># SyntaxError: unexpected EOF while parsing</span>
</span></span></code></pre></div><br><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>File <span style=color:#e6db74>&#34;&lt;ipython-input-14-dbee6e3a1c9e&gt;&#34;</span>, line <span style=color:#ae81ff>1</span>
</span></span><span style=display:flex><span>    print(<span style=color:#e6db74>&#34;Hello, World!&#34;</span>  <span style=color:#75715e># SyntaxError: unexpected EOF while parsing</span>
</span></span><span style=display:flex><span>                                                                      <span style=color:#f92672>^</span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>SyntaxError</span>: incomplete input
</span></span></code></pre></div><br><br><p><a href=https://docs.python.org/3/library/exceptions.html target=_blank>Exceptions</a> are <em><strong>runtime</strong></em> errors occurring during the execution of a Python program. They occur due to unforeseen conditions or events while the program is running. These can often be handled using try-except blocks to gracefully handle errors without crashing the program. Using our very first function as an example:</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#75715e>#Fixed code</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>def</span> <span style=color:#a6e22e>my_function</span>(a , b):
</span></span><span style=display:flex><span> print( a<span style=color:#f92672>+</span>b )
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#75715e>#Test</span>
</span></span><span style=display:flex><span>my_function(<span style=color:#ae81ff>5</span>, <span style=color:#e6db74>&#34;6&#34;</span>) <span style=color:#75715e># This will raise a TypeError</span>
</span></span></code></pre></div><br><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#f92672>---------------------------------------------------------------------------</span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>TypeError</span>                                 Traceback (most recent call last)
</span></span><span style=display:flex><span><span style=color:#f92672>&lt;</span>ipython<span style=color:#f92672>-</span>input<span style=color:#f92672>-</span><span style=color:#ae81ff>17</span><span style=color:#f92672>-</span>bf67fa8c0e95<span style=color:#f92672>&gt;</span> <span style=color:#f92672>in</span> <span style=color:#f92672>&lt;</span>cell line: <span style=color:#ae81ff>6</span><span style=color:#f92672>&gt;</span>()
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>4</span> 
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>5</span> <span style=color:#75715e>#Test</span>
</span></span><span style=display:flex><span><span style=color:#f92672>----&gt;</span> <span style=color:#ae81ff>6</span> my_function(<span style=color:#ae81ff>5</span>, <span style=color:#e6db74>&#34;6&#34;</span>)
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#f92672>&lt;</span>ipython<span style=color:#f92672>-</span>input<span style=color:#f92672>-</span><span style=color:#ae81ff>17</span><span style=color:#f92672>-</span>bf67fa8c0e95<span style=color:#f92672>&gt;</span> <span style=color:#f92672>in</span> my_function(a, b)
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>1</span> <span style=color:#75715e>#Fixed code</span>
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>2</span> <span style=color:#66d9ef>def</span> <span style=color:#a6e22e>my_function</span>(a , b):
</span></span><span style=display:flex><span><span style=color:#f92672>----&gt;</span> <span style=color:#ae81ff>3</span>  print( a<span style=color:#f92672>+</span>b )
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>4</span> 
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>5</span> <span style=color:#75715e>#Test</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>TypeError</span>: unsupported operand type(s) <span style=color:#66d9ef>for</span> <span style=color:#f92672>+</span>: <span style=color:#e6db74>&#39;int&#39;</span> <span style=color:#f92672>and</span> <span style=color:#e6db74>&#39;str&#39;</span>
</span></span></code></pre></div><p>We can observe that passing a non-numeric argument to <code>my_function</code> raises a <code>TypeError</code>, which in itself is a type of exception error.</p><br><ul><li>Division by zero:</li></ul><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>x <span style=color:#f92672>=</span> <span style=color:#ae81ff>5</span>
</span></span><span style=display:flex><span>y <span style=color:#f92672>=</span> <span style=color:#ae81ff>0</span>
</span></span><span style=display:flex><span>result <span style=color:#f92672>=</span> x <span style=color:#f92672>/</span> y  <span style=color:#75715e># ZeroDivisionError: division by zero</span>
</span></span></code></pre></div><br><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#f92672>---------------------------------------------------------------------------</span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>ZeroDivisionError</span>                         Traceback (most recent call last)
</span></span><span style=display:flex><span><span style=color:#f92672>&lt;</span>ipython<span style=color:#f92672>-</span>input<span style=color:#f92672>-</span><span style=color:#ae81ff>18</span><span style=color:#f92672>-</span><span style=color:#ae81ff>53</span>a6f18aaf2b<span style=color:#f92672>&gt;</span> <span style=color:#f92672>in</span> <span style=color:#f92672>&lt;</span>cell line: <span style=color:#ae81ff>3</span><span style=color:#f92672>&gt;</span>()
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>1</span> x <span style=color:#f92672>=</span> <span style=color:#ae81ff>5</span>
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>2</span> y <span style=color:#f92672>=</span> <span style=color:#ae81ff>0</span>
</span></span><span style=display:flex><span><span style=color:#f92672>----&gt;</span> <span style=color:#ae81ff>3</span> result <span style=color:#f92672>=</span> x <span style=color:#f92672>/</span> y  <span style=color:#75715e># ZeroDivisionError: division by zero</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>ZeroDivisionError</span>: division by zero
</span></span></code></pre></div><br><ul><li>Accessing an out-of-range index</li></ul><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>my_list <span style=color:#f92672>=</span> [<span style=color:#ae81ff>1</span>, <span style=color:#ae81ff>2</span>, <span style=color:#ae81ff>3</span>]
</span></span><span style=display:flex><span>value <span style=color:#f92672>=</span> my_list[<span style=color:#ae81ff>5</span>]  <span style=color:#75715e># IndexError: list index out of range</span>
</span></span></code></pre></div><br><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#f92672>---------------------------------------------------------------------------</span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>IndexError</span>                                Traceback (most recent call last)
</span></span><span style=display:flex><span><span style=color:#f92672>&lt;</span>ipython<span style=color:#f92672>-</span>input<span style=color:#f92672>-</span><span style=color:#ae81ff>19</span><span style=color:#f92672>-</span>c5deeffe54f1<span style=color:#f92672>&gt;</span> <span style=color:#f92672>in</span> <span style=color:#f92672>&lt;</span>cell line: <span style=color:#ae81ff>2</span><span style=color:#f92672>&gt;</span>()
</span></span><span style=display:flex><span>      <span style=color:#ae81ff>1</span> my_list <span style=color:#f92672>=</span> [<span style=color:#ae81ff>1</span>, <span style=color:#ae81ff>2</span>, <span style=color:#ae81ff>3</span>]
</span></span><span style=display:flex><span><span style=color:#f92672>----&gt;</span> <span style=color:#ae81ff>2</span> value <span style=color:#f92672>=</span> my_list[<span style=color:#ae81ff>5</span>]  <span style=color:#75715e># IndexError: list index out of range</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#a6e22e>IndexError</span>: list index out of range
</span></span></code></pre></div><br><p>The takeaway here is that exception errors will occur whenever the correct syntax of your python code yields an error.</p><br><br><h4 id=exception-handling-the-basics>Exception Handling: The Basics</h4><p>Python&rsquo;s exception handling framework consists of four distinct <strong>blocks</strong>, each strategically designed to address specific tasks within an error-handling statement.</p><br><h5 id=tryexcept-blocks>Try/Except Blocks</h5><p>This block is used to catch and handle exceptions that occur within a try block. When an exception is raised in the try block, the code within the corresponding except block is executed, allowing you to handle the exception by providing specific error-handling logic.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#66d9ef>try</span>:
</span></span><span style=display:flex><span>    <span style=color:#75715e># Code that might raise an exception</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>except</span> SomeException:
</span></span><span style=display:flex><span>    <span style=color:#75715e># Handle SomeException</span>
</span></span></code></pre></div><br><h5 id=tryfinally-blocks>Try/Finally Blocks</h5><p>The <code>finally</code> block gets executed void of a raised exception or lack thereof. We employ the <code>try/finally</code> construct in situations where we desire exceptions to propagate upwards in the call stack, yet simultaneously need to execute cleanup code, ensuring it runs even when exceptions are raised. This ensures that essential cleanup operations are consistently performed, regardless of whether an exception occurs. Common cleanup operations involve tasks like closing files or releasing resources.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>handle <span style=color:#f92672>=</span> open(<span style=color:#e6db74>&#39;somefile.txt&#39;</span>)  <span style=color:#75715e>#May raise IOError</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>try</span>:
</span></span><span style=display:flex><span>    <span style=color:#75715e># Code that might raise an exception</span>
</span></span><span style=display:flex><span>    data <span style=color:#f92672>=</span> handle<span style=color:#f92672>.</span>read()  <span style=color:#75715e>#May raise UnicodeDecodeError</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>finally</span>:
</span></span><span style=display:flex><span>    handle<span style=color:#f92672>.</span>close() <span style=color:#75715e>#Will always run after `try:`</span>
</span></span></code></pre></div><br><p>Regardless of whether an exception occurs, the finally block guarantees that the file is closed, ensuring proper cleanup and resource management. This is a common pattern for ensuring that critical resources are released, even in the presence of exceptions. To illustrate the above better, observe the code output below:</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#75715e>#handle = open(&#39;somefile.txt&#39;)  #May raise IOError</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>try</span>:
</span></span><span style=display:flex><span>    <span style=color:#75715e># Code that might raise an exception</span>
</span></span><span style=display:flex><span>    <span style=color:#75715e>#data = handle.read()  #May raise UnicodeDecodeError</span>
</span></span><span style=display:flex><span>    x <span style=color:#f92672>=</span> <span style=color:#ae81ff>10</span> <span style=color:#f92672>/</span> <span style=color:#ae81ff>0</span>  <span style=color:#75715e># This will raise a ZeroDivisionError</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>except</span> <span style=color:#a6e22e>ZeroDivisionError</span> <span style=color:#66d9ef>as</span> e:
</span></span><span style=display:flex><span>    print(<span style=color:#e6db74>f</span><span style=color:#e6db74>&#34;An error occurred: </span><span style=color:#e6db74>{</span>e<span style=color:#e6db74>}</span><span style=color:#e6db74>&#34;</span>) <span style=color:#75715e>#This will run when there is an exception</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>finally</span>:
</span></span><span style=display:flex><span>    <span style=color:#75715e>#handle.close() #Will always run after `try:`</span>
</span></span><span style=display:flex><span>    print(<span style=color:#e6db74>&#34;Cleanup: Closing open files, releasing resources, etc.&#34;</span>)
</span></span></code></pre></div><br><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>An error occurred: division by zero
</span></span><span style=display:flex><span>Cleanup: Closing open files, releasing resources, etc<span style=color:#f92672>.</span>
</span></span></code></pre></div><br><blockquote><p>A <code>try block</code> executes its code until it encounters the first exception. Within the <code>except block</code>, which serves as the exception handler, we have the ability to define the program&rsquo;s response to that specific exception. It is also possible to anticipate and handle multiple types of exceptions separately.</p></blockquote><br><h5 id=tryelse-blocks>Try/Else Blocks</h5><p>The <code>else</code> block is executed whenever there are no exceptions raised in the preceding try block.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#66d9ef>try</span>:
</span></span><span style=display:flex><span>    <span style=color:#75715e># Code that might raise an exception</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>except</span> SomeException:
</span></span><span style=display:flex><span>    <span style=color:#75715e># Handle SomeException</span>
</span></span><span style=display:flex><span><span style=color:#66d9ef>else</span>:
</span></span><span style=display:flex><span>    <span style=color:#75715e># Code to run when no exceptions occur</span>
</span></span></code></pre></div><br><p>The use of try/except/else allows us to explicitly specify which exceptions our code will handle and which exceptions will be allowed to propagate upward. This approach also aids in reducing the amount of code within the try block, resulting in improved code readability. Assuming we want to load JSON dictionary data from a string and return the value of a random key from it:</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#66d9ef>def</span> <span style=color:#a6e22e>load_json_key</span>(data, key):
</span></span><span style=display:flex><span>  <span style=color:#66d9ef>try</span>:
</span></span><span style=display:flex><span>    result_dict <span style=color:#f92672>=</span> json<span style=color:#f92672>.</span>loads(data) <span style=color:#75715e>#May raise ValueError</span>
</span></span><span style=display:flex><span>  <span style=color:#66d9ef>except</span> <span style=color:#a6e22e>ValueError</span> <span style=color:#66d9ef>as</span> e:
</span></span><span style=display:flex><span>      <span style=color:#66d9ef>raise</span> KeyEror <span style=color:#f92672>from</span> e
</span></span><span style=display:flex><span>  <span style=color:#66d9ef>else</span>:
</span></span><span style=display:flex><span>      <span style=color:#66d9ef>return</span> result_dict[key]   <span style=color:#75715e>#May raise KeyError</span>
</span></span></code></pre></div><br><p>The <code>json.loads(data)</code> function loads JSON data from the data variable. If the data isn&rsquo;t valid JSON, a <code>ValueError</code> is raised and passed up to the calling code. Within the <code>except block</code>, if a <code>ValueError</code> is caught, it&rsquo;s replaced with a <code>KeyError</code> using <code>raise KeyError from e</code>. This new KeyError then propagates up to the calling code.</p><p>If the JSON data is valid, the <code>else block</code> is executed. It attempts to access a specific key in <code>result_dict</code>. If the key isn&rsquo;t found, a <code>KeyError</code> is raised within the else block and propagates up to the calling code. The else clause serves to visually separate the code following the try/except block from the except block. This distinction enhances the clarity of the exception propagation behavior.</p><br><h4 id=taking-advantage-of-each-block-tryexceptelsefinally>Taking advantage of each block (try/except/else/finally)</h4><p>To handle it all in one compound statement, we can employ <code>try/except/else/finally</code> functionalities together. Combining all the blocks provides a comprehensive way to manage errors and control program flow. Here&rsquo;s a brief illustration:</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#66d9ef>def</span> <span style=color:#a6e22e>divide</span>(x, y):
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>try</span>:
</span></span><span style=display:flex><span>        result <span style=color:#f92672>=</span> x <span style=color:#f92672>/</span> y  <span style=color:#75715e># May raise ZeroDivisionError</span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>except</span> <span style=color:#a6e22e>ZeroDivisionError</span> <span style=color:#66d9ef>as</span> e:
</span></span><span style=display:flex><span>        print(<span style=color:#e6db74>f</span><span style=color:#e6db74>&#34;Error: </span><span style=color:#e6db74>{</span>e<span style=color:#e6db74>}</span><span style=color:#e6db74>&#34;</span>)
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>else</span>:
</span></span><span style=display:flex><span>        print(<span style=color:#e6db74>f</span><span style=color:#e6db74>&#34;Result: </span><span style=color:#e6db74>{</span>result<span style=color:#e6db74>}</span><span style=color:#e6db74>&#34;</span>)
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>finally</span>:
</span></span><span style=display:flex><span>        print(<span style=color:#e6db74>&#34;Cleanup: Closing resources&#34;</span>)
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#75715e># Test cases</span>
</span></span><span style=display:flex><span>divide(<span style=color:#ae81ff>10</span>, <span style=color:#ae81ff>2</span>)  <span style=color:#75715e># No exception, both &#39;else&#39; and &#39;finally&#39; run</span>
</span></span><span style=display:flex><span>divide(<span style=color:#ae81ff>10</span>, <span style=color:#ae81ff>0</span>)  <span style=color:#75715e># Exception (ZeroDivisionError), &#39;except&#39; and &#39;finally&#39; run</span>
</span></span></code></pre></div><br><p>Test case 1-No exception, both &rsquo;else&rsquo; and &lsquo;finally&rsquo; run:</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>Result: <span style=color:#ae81ff>5.0</span>
</span></span><span style=display:flex><span>Cleanup: Closing resources
</span></span></code></pre></div><br><p>Test case 2- Exception (ZeroDivisionError), &rsquo;except&rsquo; and &lsquo;finally&rsquo; run:</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span>Error: division by zero
</span></span><span style=display:flex><span>Cleanup: Closing resources
</span></span></code></pre></div><br><p>As we have illustrated, each block has a distinct role, and they work together intuitively:</p><p>The <code>try block</code> encloses the code that may potentially raise an exception. It&rsquo;s the starting point where we anticipate and handle potential errors. This block is essential for maintaining program stability. The <code>except block</code> comes into play when an exception is raised within the try block. It simply allows us to specify how to handle different types of exceptions. By catching specific exceptions, we can provide tailored error-handling strategies. This block enhances program resilience by gracefully dealing with errors.</p><p>The <code>else block</code> is executed only if no exceptions were raised in the try block. It allows us to place code that should run when everything runs flawless. This block is useful for separating error-handling logic from regular code, making the code more readable and maintainable. Lastly, The <code>finally block</code> is executed regardless of whether an exception was raised or not. This block ensures that all essential cleanup operations are performed, contributing to code reliability.</p><blockquote><p>The combined layout is useful as it encourages structured error handling and promotes key benefits such as:</p></blockquote><ul><li>Clarity: Each block&rsquo;s role is clearly defined, making the code easier to understand and maintain. Error-handling logic is separated from regular code, enhancing code readability.</li><li>Resilience: The except block allows for specific exception handling, enabling the program to respond appropriately to various error scenarios. This improves program resilience by addressing potential issues gracefully.</li><li>Robustness: The finally block guarantees that critical cleanup operations are executed, even in the presence of exceptions. This helps maintain the integrity of the program.</li><li>Control: By combining all these blocks, you have fine-grained control over how your program responds to errors, ensuring that it continues to function smoothly even in challenging situations.</li></ul><p>In the next post, we explore some advanced techniques coupled with best practices for gracefully handling exceptions.</p><br><br><h4 id=credits>Credits</h4><p><a href=https://storyset.com/people>People illustrations by Storyset</a> on Freepik.</p></div><div class="row ps-3 pe-3"><div class="col-md-6 share-buttons"><strong>Share on:</strong>
<a class="btn icon-button bg-facebook" href="https://www.facebook.com/sharer.php?u=https%3a%2f%2fcarlvinjerry.github.io%2fposts%2fprogramming%2fcode-sustainability%2ferror-handling%2fpython-exceptions%2f" target=_blank><i class="fab fa-facebook"></i>
</a><a class="btn icon-button bg-twitter" href="https://twitter.com/share?url=https%3a%2f%2fcarlvinjerry.github.io%2fposts%2fprogramming%2fcode-sustainability%2ferror-handling%2fpython-exceptions%2f&text=Mastering%20Python%27s%20Exceptional%20Complexity&via=R%ce%9eCURSION" target=_blank><i class="fab fa-twitter"></i>
</a><a class="btn icon-button bg-reddit" href="https://reddit.com/submit?url=https%3a%2f%2fcarlvinjerry.github.io%2fposts%2fprogramming%2fcode-sustainability%2ferror-handling%2fpython-exceptions%2f&title=Mastering%20Python%27s%20Exceptional%20Complexity" target=_blank><i class="fab fa-reddit"></i>
</a><a class="btn icon-button bg-linkedin" href="https://www.linkedin.com/shareArticle?url=https%3a%2f%2fcarlvinjerry.github.io%2fposts%2fprogramming%2fcode-sustainability%2ferror-handling%2fpython-exceptions%2f&title=Mastering%20Python%27s%20Exceptional%20Complexity" target=_blank><i class="fab fa-linkedin"></i>
</a><a class="btn icon-button bg-mastodon" href="https://mastodon.social/share?text=Mastering%20Python%27s%20Exceptional%20Complexity - https%3a%2f%2fcarlvinjerry.github.io%2fposts%2fprogramming%2fcode-sustainability%2ferror-handling%2fpython-exceptions%2f" target=_blank><i class="fab fa-mastodon"></i>
</a><a class="btn icon-button bg-whatsapp" href="https://api.whatsapp.com/send?text=Mastering%20Python%27s%20Exceptional%20Complexity https%3a%2f%2fcarlvinjerry.github.io%2fposts%2fprogramming%2fcode-sustainability%2ferror-handling%2fpython-exceptions%2f" target=_blank><i class="fab fa-whatsapp"></i>
</a><a class="btn icon-button" href="mailto:?subject=Mastering%20Python%27s%20Exceptional%20Complexity&body=https%3a%2f%2fcarlvinjerry.github.io%2fposts%2fprogramming%2fcode-sustainability%2ferror-handling%2fpython-exceptions%2f" target=_blank><i class="fas fa-envelope-open-text"></i></a></div><div class="col-md-6 btn-improve-page"><a href=https://github.com/carlvinjerry/carlvinjerry.github.io/edit/main/content/posts/programming/Code-Sustainability/Error%20Handling/python%20exceptions/index.md title="Improve this page" target=_blank rel=noopener><i class="fas fa-code-branch"></i>
Improve this page</a></div></div><hr><div class="row next-prev-navigator"><div class="col-md-6 previous-article"><a href=/posts/introduction/ title=Introduction class="btn filled-button"><div><i class="fas fa-chevron-circle-left"></i> Prev</div><div class=next-prev-text>Introduction</div></a></div><div class="col-md-6 next-article"><a href=/posts/programming/code-sustainability/succint-fsharp/currying/ title="Currying and Partial Application in F#" class="btn filled-button"><div>Next <i class="fas fa-chevron-circle-right"></i></div><div class=next-prev-text>Currying and Partial Application in F#</div></a></div></div><hr><div id=disqus_thread></div><script type=text/javascript>(function(){if(window.location.hostname=="localhost")return;var t,e=document.createElement("script");e.type="text/javascript",e.async=!0,t="i_carlvin",e.src="//"+t+".disqus.com/embed.js",(document.getElementsByTagName("head")[0]||document.getElementsByTagName("body")[0]).appendChild(e)})()</script><noscript>Please enable JavaScript to view the
<a href=https://disqus.com/?ref_noscript>comments powered by Disqus.</a></noscript><a href=https://disqus.com/ class=dsq-brlink>comments powered by <span class=logo-disqus>Disqus</span></a></div></div></div><a id=scroll-to-top class=btn type=button data-bs-toggle=tooltip data-bs-placement=left title="Scroll to top"><i class="fas fa-chevron-circle-up"></i></a></section><section class=toc-section id=toc-section><div class=toc-holder><h5 class="text-center ps-3">Table of Contents</h5><hr><div class=toc><nav id=TableOfContents><ul><li><ul><li><ul><li><a href=#introduction>Introduction</a><ul><li><a href=#exceptions-versus-syntax-errors>Exceptions versus Syntax Errors</a></li></ul></li><li><a href=#exception-handling-the-basics>Exception Handling: The Basics</a><ul><li><a href=#tryexcept-blocks>Try/Except Blocks</a></li><li><a href=#tryfinally-blocks>Try/Finally Blocks</a></li><li><a href=#tryelse-blocks>Try/Else Blocks</a></li></ul></li><li><a href=#taking-advantage-of-each-block-tryexceptelsefinally>Taking advantage of each block (try/except/else/finally)</a></li><li><a href=#credits>Credits</a></li></ul></li></ul></li></ul></nav></div></div></section></div><footer id=footer class="container-fluid text-center align-content-center footer pb-2"><div class="container pt-5"><div class="row text-start"><div class="col-md-4 col-sm-12"><h5>Navigation</h5><ul><li class=nav-item><a class=smooth-scroll href=https://carlvinjerry.github.io/#about>About</a></li><li class=nav-item><a class=smooth-scroll href=https://carlvinjerry.github.io/#recent-posts>Recent Posts</a></li><li class=nav-item><a class=smooth-scroll href=https://carlvinjerry.github.io/#projects>Projects</a></li></ul></div><div class="col-md-4 col-sm-12"><h5>Contact me:</h5><ul><li><a href=mailto:carlvinjerry@gmail.com%20%7c%20cmwange@beyondrawdata.co.ke target=_blank rel=noopener><span><i class="fas fa-envelope"></i></span> <span>carlvinjerry@gmail.com | cmwange@beyondrawdata.co.ke</span></a></li><li><a href=https://github.com/carlvinjerry target=_blank rel=noopener><span><i class="fab fa-github"></i></span> <span>carlvinjerry</span></a></li><li><a href=https://www.linkedin.com/in/carlvinmwange target=_blank rel=noopener><span><i class="fab fa-linkedin"></i></span> <span>Carlvin Jerry Mwange</span></a></li><li><a href=tel:+254%20%280%29747%20942%20588 target=_blank rel=noopener><span><i class="fas fa-phone-alt"></i></span> <span>+254 (0)747 942 588</span></a></li><li><a href=https://www.researchgate.net/profile/carlvinmwange target=_blank rel=noopener><span><i class="fab fa-researchgate"></i></span> <span>Carlvin Jerry Mwange</span></a></li></ul></div></div></div><hr><div class=container><p id=disclaimer><strong>Liability Notice:</strong> This theme is under MIT license. So, you can use it for non-commercial, commercial, or private uses. You can modify or distribute the theme without requiring any permission from the theme author. However, the theme author does not provide any warranty or takes any liability for any issue with the theme.</p></div><hr><div class=container><div class="row text-start"><div class=col-md-4><a id=theme href=https://github.com/hugo-toha/toha target=_blank rel=noopener><img src=/images/theme-logo_hu16779671404603505019.png alt="Toha Theme Logo">
Toha</a></div><div class="col-md-4 text-center">© 2020 Copyright.</div><div class="col-md-4 text-end"><a id=hugo href=https://gohugo.io/ target=_blank rel=noopener>Powered by
<img src=/images/hugo-logo.svg alt="Hugo Logo" height=18></a></div></div></div></footer><script src=/application.b1192b27bb9da1367104dd2311a23a6f78cc8e16c840ce3f8406443569df5b65.js integrity="sha256-sRkrJ7udoTZxBN0jEaI6b3jMjhbIQM4/hAZENWnfW2U=" defer></script></body></html>